### PR TITLE
[SPARK-36894][SPARK-37077][PYTHON] Synchronize RDD.toDF annotations with SparkSession and SQLContext .createDataFrame variants.

### PR DESCRIPTION
**Pull request by zero323**
### What changes were proposed in this pull request?

This pull request synchronizes `RDD.toDF` annotations with `SparkSession.createDataFrame` and `SQLContext.createDataFrame` variants.

Additionally, it fixes recent regression in `SQLContext.createDataFrame` (SPARK-37077), where `RDD` is no longer consider a valid input.

### Why are the changes needed?

- Adds support for providing `str` schema.
- Add supports for converting `RDDs` of "atomic" values, if schema is provided.

Additionally it introduces a `TypeVar` representing supported "atomic" values. This was done to avoid issue with manual data tests, where the following

```python
sc.parallelize([1]).toDF(schema=IntegerType())
```

results in 

```
error: No overload variant of "toDF" of "RDD" matches argument type "IntegerType"  [call-overload]
note: Possible overload variants:
note:     def toDF(self, schema: Union[List[str], Tuple[str, ...], None] = ..., sampleRatio: Optional[float] = ...) -> DataFrame
note:     def toDF(self, schema: Union[StructType, str, None] = ...) -> DataFrame
```

when `Union` type is used (this problem doesn't surface when non-self bound is used).

### Does this PR introduce _any_ user-facing change?

Type checker only.

Please note, that these annotations serve primarily to support documentation, as checks on `self` types are still very limited.

### How was this patch tested?

Existing tests and manual data tests.

### CODE REVIEW

**Comment by zero323** [comment]
FYI @ueshin

**Comment by zero323** [comment]
__Note__

I am going to keep it as draft for now, because I realized we prefer to extend the scope of the ticket to synchronize `RDD.toDF` with `SparkSession.createDataFrame`.

File: python/pyspark/rdd.pyi [hunk 1]
```diff
@@ -445,11 +445,18 @@ class RDD(Generic[T]):
...
         sampleRatio: Optional[float] = ...,
     ) -> DataFrame: ...
     @overload
-    def toDF(self: RDD[RowLike], schema: Optional[StructType] = ...) -> DataFrame: ...
+    def toDF(
+        self: RDD[RowLike], schema: Optional[Union[StructType, str]] = ...
+    ) -> DataFrame: ...
+    @overload
+    def toDF(
+        self: RDD[Union[DateTimeLiteral, LiteralType, DecimalLiteral]],
```
Comment (zero323) [inline python/pyspark/rdd.pyi:457]: It seems like this part will requires some changes (switch to `TypeVar`). I'll investigate this and provide an update.

**Comment by zero323** [comment]
__Note__:

This variant is dropped, because it is non-functional:

https://github.com/apache/spark/blob/722ac1b8b7f86fdeedf20cc11c7f547e7038029c/python/pyspark/sql/session.py#L636-L642

If only names are provided, we infer schema and there is no verification.

**Review by HyukjinKwon — APPROVED** [review]

**Comment by zero323** [comment]
Merged to master.

**Comment by sarutak** [comment]
@zero323 Sorry but I'll revert this because of Python linter failure on GA.
I've retried but it still failed.
Could you check it out?
https://github.com/apache/spark/runs/4093063784?check_suite_focus=true#step:14:50

**Comment by zero323** [comment]
Thanks @sarutak,  on it.

**Comment by HyukjinKwon** [comment]
Thanks guys.

**Pull request closed without merging**